govcd/vapp_network.go

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2020 VMware, Inc.  All rights reserved.  Licensed under the Apache v2 License.
+ */
+
+package govcd
+
+import (
+	"fmt"
+	"github.com/vmware/go-vcloud-director/v2/types/v56"
+	"github.com/vmware/go-vcloud-director/v2/util"
+	"net/http"
+)
+
+// UpdateNetworkFirewallRules updates vApp networks firewall rules.
+// Returns pointer to types.VAppNetwork or error
+func (vapp *VApp) UpdateNetworkFirewallRules(networkId string, firewallRules []*types.FirewallRule, defaultAction string, logDefaultAction bool) (*types.VAppNetwork, error) {
+	task, err := vapp.UpdateNetworkFirewallRulesAsync(networkId, firewallRules, defaultAction, logDefaultAction)
+	if err != nil {
+		return nil, err
+	}
+	err = task.WaitTaskCompletion()
+	if err != nil {
+		return nil, fmt.Errorf("%s", combinedTaskErrorMessage(task.Task, err))
+	}
+
+	return vapp.GetVappNetworkById(networkId, false)
+}
+
+// UpdateNetworkFirewallRulesAsync asynchronously updates vApp networks firewall rules.
+// Returns task or error
+func (vapp *VApp) UpdateNetworkFirewallRulesAsync(networkId string, firewallRules []*types.FirewallRule, defaultAction string, logDefaultAction bool) (Task, error) {
+	util.Logger.Printf("[TRACE] UpdateNetworkFirewallRulesAsync with values: id: %s and firewallServiceConfiguration: %#v", networkId, firewallRules)
+	uuid := extractUuid(networkId)
+	networkToUpdate, err := vapp.GetVappNetworkById(uuid, true)
+	if err != nil {
+		return Task{}, err
+	}
+
+	if networkToUpdate.Configuration.Features == nil {
+		networkToUpdate.Configuration.Features = &types.NetworkFeatures{}
+	}
+	networkToUpdate.Xmlns = types.XMLNamespaceVCloud
+
+	if networkToUpdate.Configuration.Features.FirewallService == nil {
+		return Task{}, fmt.Errorf("provided network isn't connecd to org network or isn't fenced")
```

Comment:
This error looks to be misleading based on what it checks. Is it incorrect or me not understanding it?

Reply:
It is `helper`. If API didn't return Firewall service XML part, that means vapp network isn't connected to org network or not fenced. In other words there isn't firewall when you connected directly ir isolated.

Reply:
Alright. Can you just put comment above (pretty much the sentence you wrote here. I think next time looking at it we will forget (the ones who didn't write it).

Reply:
done